Linter "testify" is broken and testify output breaks linter "test"

When using `testify` linter together with [testify](https://github.com/stretchr/testify), gometalinter fails to detect errors

Given the following test file:
```go
package test

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestHello(t *testing.T) {
	assert.Equal(t, "This is not", "equal to this")
	t.Fatal("this should fail")
}
```

When I run `go test ./...` I get the following output showing errors on line `10` and `11`:
```
--- FAIL: TestHello (0.00s)
	other_test.go:10: 
			Error Trace:	other_test.go:10
			Error:      	Not equal: 
			            	expected: "This is not"
			            	actual  : "equal to this"
			            	
			            	Diff:
			            	--- Expected
			            	+++ Actual
			            	@@ -1 +1 @@
			            	-This is not
			            	+equal to this
			Test:       	TestHello
	other_test.go:11: this should fail
FAIL
FAIL	test	0.003s
```
However when I run gometalinter with linters `testìfy` and `test` I only get an error from `test` linter for line `10` without a description:
```
other_test.go:10::error:  (test)
```
Debug output tells me that the patterns fails to match the output:
```
...
DEBUG: [Jun 15 09:40:02.093] [test.1]: executing /usr/local/go/bin/go test .
DEBUG: [Jun 15 09:40:02.094] [testify.2]: executing /usr/local/go/bin/go test .
DEBUG: [Jun 15 09:40:02.776] [testify.2]: warning: /usr/local/go/bin/go returned exit status 1: --- FAIL: TestHello (0.00s)
	other_test.go:10: 
			Error Trace:	other_test.go:10
			Error:      	Not equal: 
			            	expected: "This is not"
			            	actual  : "equal to this"
			            	
			            	Diff:
			            	--- Expected
			            	+++ Actual
			            	@@ -1 +1 @@
			            	-This is not
			            	+equal to this
			Test:       	TestHello
	other_test.go:11: this should fail
FAIL
FAIL	test	0.003s

DEBUG: [Jun 15 09:40:02.776] [testify.2]: testify hits 0: Location:\s+(?P<path>.*?\.go):(?P<line>\d+)$\s+Error:\s+(?P<message>[^\n]+)
DEBUG: [Jun 15 09:40:02.776] [testify.2]: testify linter took 682.592608ms
DEBUG: [Jun 15 09:40:02.808] [test.1]: warning: /usr/local/go/bin/go returned exit status 1: --- FAIL: TestHello (0.00s)
	other_test.go:10: 
			Error Trace:	other_test.go:10
			Error:      	Not equal: 
			            	expected: "This is not"
			            	actual  : "equal to this"
			            	
			            	Diff:
			            	--- Expected
			            	+++ Actual
			            	@@ -1 +1 @@
			            	-This is not
			            	+equal to this
			Test:       	TestHello
	other_test.go:11: this should fail
FAIL
FAIL	test	0.003s

DEBUG: [Jun 15 09:40:02.808] [test.1]: test hits 1: ^--- FAIL: .*$\s+(?P<path>.*?\.go):(?P<line>\d+): (?P<message>.*)$
DEBUG: [Jun 15 09:40:02.808] [test.1]: test linter took 714.534469ms
DEBUG: [Jun 15 09:40:02.808] nolint: parsing other_test.go for directives
DEBUG: [Jun 15 09:40:02.808] nolint: parsing other_test.go took 148.701µs
other_test.go:10::error:  (test)
DEBUG: [Jun 15 09:40:02.808] total elapsed time 718.915311ms
```
Looks like the pattern for `testify` is outdated and pattern for `test` could be improved

--- a/linters.go
+++ b/linters.go
@@ -366,7 +366,7 @@ var defaultLinters = map[string]LinterConfig{
 	},
 	"testify": {
 		Command:           `go test`,
-		Pattern:           `Location:\s+(?P<path>.*?\.go):(?P<line>\d+)$\s+Error:\s+(?P<message>[^\n]+)`,
+		Pattern:           `(?m:^\s+Error Trace:\s+(?P<path>.+?.go):(?P<line>\d+)\n\s+Error:\s+(?P<message>.+?)[:\s]*$)`,
 		PartitionStrategy: partitionPathsAsPackages,
 	},
 	"unconvert": {

--- a/linters.go
+++ b/linters.go
@@ -361,7 +361,7 @@ var defaultLinters = map[string]LinterConfig{
 	},
 	"test": {
 		Command:           `go test`,
-		Pattern:           `^--- FAIL: .*$\s+(?P<path>.*?\.go):(?P<line>\d+): (?P<message>.*)$`,
+		Pattern:           `(?m:^\t(?P<path>.*?\.go):(?P<line>\d+): (?P<message>.+)$)`,
 		PartitionStrategy: partitionPathsAsPackages,
 	},
 	"testify": {

--- a/linters_test.go
+++ b/linters_test.go
@@ -2,6 +2,7 @@ package main
 
 import (
 	"reflect"
+	"regexp"
 	"runtime"
 	"testing"
 
@@ -54,6 +55,70 @@ func TestValidateLinters(t *testing.T) {
 	}
 	err = validateLinters(lintersFromConfig(config), config)
 	require.NoError(t, err)
+}
+
+func TestLinter_test(t *testing.T) {
+	exampleOutput := `--- FAIL: TestHello (0.00s)
+	other_test.go:11: 
+			Error Trace:	other_test.go:11
+			Error:      	Not equal: 
+			            	expected: "This is not"
+			            	actual  : "equal to this"
+			            	
+			            	Diff:
+			            	--- Expected
+			            	+++ Actual
+			            	@@ -1 +1 @@
+			            	-This is not
+			            	+equal to this
+			Test:       	TestHello
+	other_test.go:12: this should fail
+	other_test.go:13: fail again
+	other_test.go:14: last fail
+	other_test.go:15:   
+	other_test.go:16: 
+	require.go:1159: 
+			Error Trace:	other_test.go:17
+			Error:      	Should be true
+			Test:       	TestHello
+FAIL
+FAIL	test	0.003s`
+
+	pattern := regexp.MustCompile(defaultLinters["test"].Pattern)
+	matches := pattern.FindAllStringSubmatch(exampleOutput, -1)
+	var errors []map[string]string
+	for _, match := range matches {
+		m := make(map[string]string)
+		for i, name := range pattern.SubexpNames() {
+			if i != 0 && name != "" {
+				m[name] = string(match[i])
+			}
+		}
+		errors = append(errors, m)
+	}
+
+	// Assert expected errors
+	assert.Equal(t, "other_test.go", errors[0]["path"])
+	assert.Equal(t, "12", errors[0]["line"])
+	assert.Equal(t, "this should fail", errors[0]["message"])
+
+	assert.Equal(t, "other_test.go", errors[1]["path"])
+	assert.Equal(t, "13", errors[1]["line"])
+	assert.Equal(t, "fail again", errors[1]["message"])
+
+	assert.Equal(t, "other_test.go", errors[2]["path"])
+	assert.Equal(t, "14", errors[2]["line"])
+	assert.Equal(t, "last fail", errors[2]["message"])
+
+	assert.Equal(t, "other_test.go", errors[3]["path"])
+	assert.Equal(t, "15", errors[3]["line"])
+	assert.Equal(t, "  ", errors[3]["message"])
+
+	// Go metalinter does not support errors without a message as there is little or no output to parse
+	// E.g. t.Fail() or t.Error("")
+	//  assert.Equal(t, "other_test.go", errors[5]["path"])
+	//	assert.Equal(t, "15", errors[5]["line"])
+	//	assert.Equal(t, "", errors[5]["message"])
 }
 
 func functionName(i interface{}) string {

--- a/linters_test.go
+++ b/linters_test.go
@@ -121,6 +121,56 @@ FAIL	test	0.003s`
 	//	assert.Equal(t, "", errors[5]["message"])
 }
 
+func TestLinter_testify(t *testing.T) {
+	exampleOutput := `--- FAIL: TestHello (0.00s)
+	other_test.go:11: 
+			Error Trace:	other_test.go:11
+			Error:      	Not equal: 
+			            	expected: "This is not"
+			            	actual  : "equal to this"
+			            	
+			            	Diff:
+			            	--- Expected
+			            	+++ Actual
+			            	@@ -1 +1 @@
+			            	-This is not
+			            	+equal to this
+			Test:       	TestHello
+	other_test.go:12: this should fail
+	other_test.go:13: fail again
+	other_test.go:14: last fail
+	other_test.go:15:   
+	other_test.go:16: 
+	require.go:1159: 
+			Error Trace:	other_test.go:17
+			Error:      	Should be true
+			Test:       	TestHello
+FAIL
+FAIL	test	0.003s`
+
+	pattern := regexp.MustCompile(defaultLinters["testify"].Pattern)
+	matches := pattern.FindAllStringSubmatch(exampleOutput, -1)
+	var errors []map[string]string
+	for _, match := range matches {
+		m := make(map[string]string)
+		for i, name := range pattern.SubexpNames() {
+			if i != 0 && name != "" {
+				m[name] = string(match[i])
+			}
+		}
+		errors = append(errors, m)
+	}
+
+	// Assert expected errors
+	assert.Equal(t, "other_test.go", errors[0]["path"])
+	assert.Equal(t, "11", errors[0]["line"])
+	assert.Equal(t, "Not equal", errors[0]["message"])
+
+	assert.Equal(t, "other_test.go", errors[1]["path"])
+	assert.Equal(t, "17", errors[1]["line"])
+	assert.Equal(t, "Should be true", errors[1]["message"])
+}
+
 func functionName(i interface{}) string {
 	return runtime.FuncForPC(reflect.ValueOf(i).Pointer()).Name()
 }

